### PR TITLE
fix(agent): align beam default supportType and improve LLM test CI

### LINKED ISSUE #118: test: cover agent-runtime, agent-tools, and Python skill modules
### Background / 背景

Follows up on #42. A coverage audit of the current codebase shows that several critical subsystems have **zero dedicated tests** despite containing core business logic. This issue tracks the highest-priority gaps.

跟进 #42。对当前代码库的覆盖率审计表明，若干关键子系统虽然包含核心业务逻辑，却完全没有专项测试。本 issue 追踪优先级最高的覆盖盲区。

### Current State / 现状

| Module | Source Files | Test Files | Status |
|--------|:---:|:---:|--------|
| `backend/src/agent-runtime/` | 19 TS | 0 | No coverage |
| `backend/src/agent-tools/` | 9 TS | 0 | No coverage |
| `backend/src/agent-skills/analysis/` | ~21 Py | 0 | No coverage |
| `backend/src/agent-skills/code-check/` | ~7 Py | 0 | No coverage |
| `backend/src/agent-skills/data-input/converters/` | ~6 Py | 0 | No coverage |
| `backend/src/services/` (untested subset) | ~18 TS | 0 | No coverage |
| `backend/src/api/` (untested routes) | 6 TS | 0 | No coverage |
| `backend/src/agent-skills/material/` | 2 Py | 0 | No coverage |

### Tasks / 任务清单

**Phase 1 — Agent runtime and tools (TypeScript)**
- [ ] Add unit tests for `backend/src/agent-runtime/` (registry, loader, executor, manifest-schema, tool-registry)
- [ ] Add unit tests for `backend/src/agent-tools/builtin/` (draft-model, validate-model, run-analysis, run-code-check, convert-model, generate-report, update-model)

**Phase 2 — Python analysis and code-check modules**
- [ ] Add unit tests for `backend/src/agent-skills/analysis/` (OpenSees static/dynamic/seismic/nonlinear adapters, simplified engines, runtime)
- [ ] Add unit tests for `backend/src/agent-skills/code-check/` (GB50009, GB50010, GB50011, GB50017, JGJ3 runners)
- [ ] Add unit tests for `backend/src/agent-skills/data-input/converters/` (model conversion pipelines)
- [ ] Add unit tests for `backend/src/agent-skills/material/` (concrete, steel material definitions)

**Phase 3 — Services and API routes (TypeScript)**
- [ ] Add unit tests for untested `backend/src/services/` files (agent-capability, agent-context, agent-session, agent-router, agent-policy, agent-validation, code-check-execution, structure-protocol-execution)
- [ ] Add integration tests for untested `backend/src/api/` routes (agent, analysis, analysis-engine, chat)

### Acceptance Criteria / 验收标准

- Every module listed above has at least one test file covering its main entry point and error paths
- New tests follow existing conventions (Jest `.test.mjs` for backend TS, `unittest` for Python)
- All new tests pass in CI without flakiness
- No regression in existing test suites

- 上述每个模块至少有一个测试文件，覆盖主入口和错误路径
- 新测试遵循现有约定（后端 TS 用 Jest `.test.mjs`，Python 用 `unittest`）
- 所有新测试在 CI 中稳定通过
- 不引入已有测试套件的回归

### PR DESCRIPTION
**Pull request by guyi2000**
## Summary
- **Fix `beam-clarify-zh` test failure**: `draft.md` instructed the LLM to default to `"cantilever"` when support type is unspecified, but `handler.ts`, `executor.ts` prompt, and test fixtures all expect `"simply-supported"`.
- **Improve LLM integration CI**: trigger reply, post-run result reporting, permission scoping, LLM call logging, and pipeline reliability fixes.

## Changes

| Commit | Description |
|--------|-------------|
| `7554be8` | Align `draft.md` default `supportType` from `"cantilever"` to `"simply-supported"` |
| `e91d88b` | Add bilingual (zh/en) text to the default instruction line |
| `8e77881` | Add trigger reply comment with action link on `/test-llm` |
| `507ade8` | Post test results (pass/fail + summary) as PR comment after run completes |
| `49d6c20` | Scope `issues: write` to job level; add `await` to async comment calls |
| `5e9e6ac` | Enable LLM call logging: fix log directory resolution (`__dirname` instead of `cwd`), add logging wrapper to test client, add `LLM_LOG_DIR` env var |
| `9b9c7e8` | Add `set -o pipefail` for correct exit code propagation; add `safeStringify` to prevent non-serializable values from crashing logging |

## Test plan
- [x] Gemini review comment addressed (bilingual consistency)
- [x] Copilot review comments addressed (permissions, await, pipefail, serialization)
- [ ] Verify `beam-clarify-zh` passes in LLM Integration CI (`/test-llm clarification`)
- [ ] Confirm trigger reply and result reporting work via `/test-llm` on a PR
- [ ] Confirm `llm-calls.jsonl` artifact is uploaded after CI run

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/src/agent-skills/structure-type/beam/draft.md [hunk 1]
```diff
@@ -15,7 +15,7 @@
 - "简支" / "simply supported" / "简支梁" → `"supportType": "simply-supported"`
 - "悬臂" / "cantilever" / "悬臂梁" → `"supportType": "cantilever"`
 - "两端固" / "fixed-fixed" → `"supportType": "fixed-fixed"`
-- 未指定支座类型时默认 `"supportType": "cantilever"`
+- 未指定支座类型时默认 `"supportType": "simply-supported"`
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/structure-type/beam/draft.md]: Good catch. Fixed in e91d88b — added the English translation to keep bilingual consistency with the rest of the file.

File: .github/workflows/llm-integration.yml [hunk 2]
```diff
@@ -29,6 +29,7 @@ on:
 permissions:
   contents: read
   pull-requests: read
+  issues: write
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml]: Good catch. Fixed in 49d6c20 — moved `issues: write` from workflow level to job-level permissions. Both `check-permission` and `llm-integration` now explicitly declare only the permissions they need.

File: .github/workflows/llm-integration.yml [hunk 3]
```diff
@@ -63,6 +64,28 @@ jobs:
...
+              ``,
+              `🚀 **[View Action Run Details](${runUrl})**`,
+            ].join('\n');
+
+            github.rest.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: body
+            })
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml]: Fixed in 49d6c20 — added `await` to both `github.rest.issues.createComment` calls.

File: .github/workflows/llm-integration.yml [hunk 4]
```diff
@@ -29,6 +29,7 @@ on:
 permissions:
   contents: read
   pull-requests: read
+  issues: write
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml:31]: PR description has already been updated to reflect all changes (draft.md fix + bilingual consistency + CI workflow improvements). No further action needed.

File: .github/workflows/llm-integration.yml [hunk 5]
```diff
@@ -81,21 +115,68 @@ jobs:
...
           DATABASE_URL: "file:.runtime/data/structureclaw-llm-ci.db"
         run: |
           FILTER_ARG="${{ needs.check-permission.outputs.filter_arg }}"
           if [ -n "$FILTER_ARG" ]; then
-            node tests/runner.mjs llm-integration --filter "$FILTER_ARG"
+            node tests/runner.mjs llm-integration --filter "$FILTER_ARG" 2>&1 | tee test-output.txt
           else
-            node tests/runner.mjs llm-integration
+            node tests/runner.mjs llm-integration 2>&1 | tee test-output.txt
           fi
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml:135]: Good catch. Fixed in 9b9c7e8 — added `set -o pipefail` before the tee pipeline so node exit codes propagate correctly.

File: .github/workflows/llm-integration.yml [hunk 6]
```diff
@@ -58,13 +63,42 @@ jobs:
...
+              repo: context.repo.repo,
+              body: body
+            })
+
   llm-integration:
     needs: check-permission
+    permissions:
+      contents: read
+      issues: write
     environment: test
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml:103]: Acknowledged the risk, but splitting into a separate job/workflow would add significant complexity for this repo's scale. The `llm-integration` job is already gated behind the `environment: test` approval gate (required for external PRs), and the CI is only triggered by authorized users (`guyi2000`, `qinsz01`). The PR code is also explicitly checked out at `needs.check-permission.outputs.pr_sha` which has already been validated by the permission check. For now, keeping the current structure is a reasonable tradeoff. If the repo grows to accept community PRs with less restrictive permissions, this should be revisited.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 7]
```diff
@@ -32,6 +36,70 @@ function createRealLlmClient(context, temperature = 0) {
...
+  const originalInvoke = model.invoke.bind(model);
+
+  model.invoke = async function (input, options) {
+    const promptStr = typeof input === "string" ? input : JSON.stringify(input);
+    const start = Date.now();
+    try {
+      const result = await originalInvoke(input, options);
+      const content = typeof result.content === "string"
+        ? result.content
+        : JSON.stringify(result.content);
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs]: Good catch. Fixed in 9b9c7e8 — wrapped all serialization in `safeStringify()` with try/catch fallback to `String()`, and added `writeLogEntry()` wrapper so log write failures are also non-blocking.

File: backend/src/utils/llm-logger.ts [hunk 8]
```diff
@@ -34,8 +37,9 @@ class LlmCallLogger {
     }
 
     try {
+      // __dirname = backend/dist/utils/ (compiled) → resolve to repo-root/.runtime/logs
       const dir = config.llmLogDir
-        || path.resolve(process.cwd(), '../.runtime/logs');
+        || path.resolve(__dirname, '../../../.runtime/logs');
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/utils/llm-logger.ts:42]: The path is actually correct. From `backend/dist/utils/`, `../../../` goes: `utils/` → `dist/` → `backend/` → repo root. Verified:

```
__dirname = /home/repo/backend/dist/utils
path.resolve(__dirname, '../../../.runtime/logs') = /home/repo/.runtime/logs
```

So the 3 `..` segments correctly reach the repo root, not `backend/`. No change needed.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 9]
```diff
@@ -32,6 +36,84 @@ function createRealLlmClient(context, temperature = 0) {
...
+ */
+let _logStream = null;
+function ensureLogStream(rootDir) {
+  if (_logStream) return _logStream;
+  if (process.env.LLM_LOG_ENABLED === "false") return null;
+  try {
+    const dir = process.env.LLM_LOG_DIR || path.join(rootDir, ".runtime", "logs");
+    fs.mkdirSync(dir, { recursive: true });
+    _logStream = fs.createWriteStream(path.join(dir, "llm-calls.jsonl"), { flags: "a" });
+    return _logStream;
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs:58]: Fixed in 222fb78 — added `stream.on('error', ...)` handler that disables further logging on error, matching the pattern in `backend/src/utils/llm-logger.ts`.

File: .github/workflows/llm-integration.yml [hunk 10]
```diff
@@ -81,21 +115,69 @@ jobs:
...
+      - name: Report results on PR
+        if: always() && needs.check-permission.outputs.trigger_issue_number != ''
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const issueNumber = Number('${{ needs.check-permission.outputs.trigger_issue_number }}');
+            const commentUser = '${{ needs.check-permission.outputs.trigger_comment_user }}';
+            const success = '${{ steps.run-tests.outcome }}' === 'success';
+            const summary = `${{ steps.summary.outputs.summary }}`;
+            const runUrl = `https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId}`;
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml:159]: Fixed in 2c34a80 — replaced template literal injection with `${{ toJson(steps.summary.outputs.summary) }}` to safely escape special characters.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 11]
```diff
@@ -32,6 +36,88 @@ function createRealLlmClient(context, temperature = 0) {
...
+function ensureLogStream(rootDir) {
+  if (_logDisabled) return null;
+  if (_logStream) return _logStream;
+  if (process.env.LLM_LOG_ENABLED === "false") { _logDisabled = true; return null; }
+  try {
+    const dir = process.env.LLM_LOG_DIR || path.join(rootDir, ".runtime", "logs");
+    fs.mkdirSync(dir, { recursive: true });
+    _logStream = fs.createWriteStream(path.join(dir, "llm-calls.jsonl"), { flags: "a" });
+    _logStream.on("error", () => { _logDisabled = true; _logStream = null; });
+    return _logStream;
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs:58]: Fixed in 2c34a80 — test wrapper now writes to `llm-calls-test.jsonl` (separate from backend's `llm-calls.jsonl`). Artifact upload updated to capture `*.jsonl` glob.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 12]
```diff
@@ -1,11 +1,15 @@
...
 const path = require("node:path");
 
 /**
  * Create a real LLM client using the backend's @langchain/openai dependency.
  * Uses `apiKey` (v1.x) parameter name. Reads config from process.env which
  * must be set before calling this function.
  *
+ * The returned client is wrapped with LLM call logging so that every
+ * invoke() call is recorded to .runtime/logs/llm-calls.jsonl.
+ *
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs:12]: Fixed in 8bb9791 — updated docstring to reference `llm-calls-test.jsonl`.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 13]
```diff
@@ -32,6 +36,88 @@ function createRealLlmClient(context, temperature = 0) {
     },
   });
 
+  return wrapWithLogging(model, context);
+}
+
+/**
+ * Self-contained LLM call logger. Writes one JSON line per invoke() call
+ * to <rootDir>/.runtime/logs/llm-calls.jsonl — same format as the backend's
+ * LlmCallLogger so the CI artifact upload picks it up automatically.
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs:45]: Fixed in 8bb9791 — updated block comment to reference `llm-calls-test.jsonl`.

File: tests/llm-integration/lib/real-llm-client.cjs [hunk 14]
```diff
@@ -32,6 +36,88 @@ function createRealLlmClient(context, temperature = 0) {
...
+
+  const modelName = context.env.LLM_MODEL || process.env.LLM_MODEL || "unknown";
+  const originalInvoke = model.invoke.bind(model);
+
+  function safeStringify(val) {
+    try {
+      return typeof val === "string" ? val : JSON.stringify(val);
+    } catch {
+      return String(val);
+    }
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply tests/llm-integration/lib/real-llm-client.cjs:79]: Good catch. Fixed in 8bb9791 — `safeStringify` now checks for `undefined` return from `JSON.stringify` and falls back to `String(val)`.

File: .github/workflows/llm-integration.yml [hunk 15]
```diff
@@ -58,13 +63,42 @@ jobs:
             COMMENT="${{ github.event.comment.body }}"
             FILTER=$(echo "$COMMENT" | sed -n 's|/test-llm[[:space:]]*\(.*\)|\1|p' | xargs)
             echo "filter_arg=$FILTER" >> "$GITHUB_OUTPUT"
+            echo "issue_number=${{ github.event.issue.number }}" >> "$GITHUB_OUTPUT"
+            echo "comment_user=${{ github.event.comment.user.login }}" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply .github/workflows/llm-integration.yml]: Good catch. Fixed in 8bb9791 — comment body is now passed via `env:` block (`COMMENT_BODY`) instead of direct bash interpolation, preventing shell injection.

File: backend/tests/llm-logger-coverage.test.mjs [hunk 16]
```diff
@@ -539,9 +539,14 @@ describe('llmCallLogger.log edge cases', () => {
...
-      await new Promise((resolve) => setTimeout(resolve, 200));
-
-      const lines = readJsonlLines(path.join(tmpDir, 'llm-calls.jsonl'));
+      // Wait for WriteStream to flush; retry on slow CI (especially Windows).
+      const jsonlPath = path.join(tmpDir, 'llm-calls.jsonl');
+      let lines = [];
+      for (let attempt = 0; attempt < 10; attempt++) {
+        await new Promise((resolve) => setTimeout(resolve, 100));
+        lines = readJsonlLines(jsonlPath);
+        if (lines.length >= 50) break;
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/tests/llm-logger-coverage.test.mjs:552]: Fixed in 9bd88ec — wrapped readJsonlLines in try/catch inside the polling loop, returning empty array on ENOENT or other read errors.

**Review by qinsz01 — APPROVED** [review]
